roxie/udplib/udptrr.cpp

```diff
@@ -136,44 +148,74 @@ class CReceiveManager : implements IReceiveManager, public CInterface
             timeouts = 0;
         }
 
-        inline bool retryOnTimeout()
+        bool canSendAny() const
         {
-            ++timeouts;
-            if (udpTraceLevel)
+            // We can send some if (a) the first available new packet is less than TRACKER_BITS above the first unreceived packet or
+            // (b) we are assuming arrival in order, and there are some marked seen that are > first unseen OR
+            // (c) the oldest in-flight packet has expired
+            if (packetsSeen.canRecord(sendSeq))
+                return true;
+            if (udpAssumeSequential && packetsSeen.hasGaps())
+                return true;
+            if (msTick()-requestTime > udpResendTimeout)
+                return true;
+            return false;
+        }
+
+        void acknowledgeRequest(const IpAddress &returnAddress, unsigned _flowSeq, unsigned _sendSeq)
+        {
+            if (flowSeq==_flowSeq)
             {
-                StringBuffer s;
-                DBGLOG("Timed out %d times waiting for send_done from %s", timeouts, dest.getIpText(s).str());
+                // It's a duplicate request-to-send - ignore it? Or assume it means they lost our ok--to-send ? MORE - probably depends on state
+                //if (udpTraceLevel || udpTraceFlow)
+                    DBGLOG("UdpFlow: ignoring duplicate reqestToSend %" SEQF "u", _flowSeq);
+                return;
             }
-            if (udpMaxRetryTimedoutReqs && (timeouts >= udpMaxRetryTimedoutReqs))
+            flowSeq = _flowSeq;
+            sendSeq = _sendSeq;
+            requestTime = msTick();
+            try
             {
-                if (udpTraceLevel)
-                    DBGLOG("Abandoning");
-                timeouts = 0;
-                return false;
+                UdpPermitToSendMsg msg;
+                msg.cmd = flowType::request_received;
+                msg.flowSeq = _flowSeq;
+                msg.destNode = returnAddress;
+                msg.max_data = 0;
+                msg.seen = packetsSeen.copy();
+                if (udpTraceLevel > 3 || udpTraceFlow)
+                {
+                    StringBuffer ipStr;
+                    DBGLOG("UdpReceiver: sending request_received msg seq %u to node=%s", _flowSeq, dest.getIpText(ipStr).str());
+                }
+                flowSocket->write(&msg, sizeof(UdpPermitToSendMsg));
+                flowPermitsSent++;
```

Comment:
I think only incremented on a single trhead, so may want to perform a more efficient increment.

Reply:
Separate Jira I think